src/sagemaker_sklearn_extension/feature_extraction/sequences.py

```diff
@@ -0,0 +1,414 @@
+
+
+class TimeSeriesFeatureExtractor(BaseEstimator, TransformerMixin):
+    """Wrap TSFlattener and TSFreshFeatureExtractor to extract time series features from multiple sequence columns.
+
+    The input is an array where rows are observations and columns are sequence features. Each sequence feature is a
+    string containing a sequence of comma-separate numbers.
+
+    For each column, TSFlattener extracts numerical values from the strings and returns a list of np.arrays as output,
+    and then TSFreshFeatureExtractor extracts time series features from each list. The outputs from each column are then
+    stacked horizontally into a single array.
+
+    See TSFlattener and TSFreshFeatureExtractor for more details.
+
+    Parameters
+    ----------
+    max_allowed_length : int (default = 10000)
+        Maximum allowed length of an input sequence. If the length of a sequence is greater than ``max_allowed_length``,
+        the transformer will truncate its beginning or end as indicated by ``trim_beginning``.
+
+    trim_beginning : bool (default = True)
+        If a sequence length exceeds ``max_allowed_length``, trim its start and only keep the last `max_allowed_length``
+        values if ``trim_beginning`` = True, otherwise trim the end and only keep the first max_allowed_length values.
+
+    augment : boolean (default=True):
+        Whether to append the tsfresh features to the original data (if True),
+        or output only the extracted tsfresh features (if False).
+        If True, also pad shorter sequences (if any) in the original data with np.nans, so that all sequences
+        match the length of the longest sequence, and interpolate them as indicated by ``interpolation_method``.
+
+    interpolation_method : {'linear', 'fill', 'zeroes', 'hybrid', None} (default='hybrid')
+        'linear': linear interpolation
+        'fill': forward fill to complete the sequences; then, backfill in case of NaNs at the start of the sequence.
+        'zeroes': pad with zeroes
+        'hybrid': replace with zeroes any NaNs at the end or start of the sequences, and forward fill NaNs in between
+         None: no interpolation
+
+    extraction_type : {'minimal', 'efficient', 'all'} (default='efficient')
+        Control the number of features extracted from tsfresh.
+        'minimal': most of the feature extractors are disabled and only a small subset is used
+        'efficient': extract 781 tsfresh features, namely all of them except for the expensive to compute ones
+        'all': extract all 787 tsfresh features
+
+    Examples
+    --------
+    >>> from sagemaker_sklearn_extension.feature_extraction.sequences import TimeSeriesFeatureExtractor
+    >>> import numpy as np
+    >>> data = [["1,, 3, 44", "3, 4, 5, 6"], ["11, 111", "1, 1, 2, 2"], ["NaN, , 1, NaN", "2, 1, 3, 2"]]
+    >>> ts_pipeline = TimeSeriesFeatureExtractor(augment=True)
+    >>> X = ts_pipeline.fit_transform(data)
+    >>> print(X.shape)
+    (3, 1570)
```

Comment:
Good point. With augment = False (the new default), we discard the original sequence and only keep the tsfresh features. This is a fixed number (781 features per sequence column) that we know in advance, which partly mitigates the risk. For instance, if the input are sequences of 50K values each, the output will still be 781 columns.

Reply:
What's the average number of sequences length in the UCR datasets you looked at? If it's closer to 2 and not 50k, we could have an expansion factor issue here and might need to be careful which feature engineering pipelines are created.

Reply:
The mean sequence length in the UCR data is 537, so on average we go from 537 to 781. 

To mitigate the expansion in case of very short sequences, we also have the option of setting extraction_type = 'minimal', which only outputs 8 features. While 'minimal' can't be the default as it gave 10% worse RED on average across all UCR data, it can be used to mitigate the expansion factor for very short sequences.